File: round1/146-+/LRUCache.py
/*
    这道题map存储节点的上一个时需要注意，修改删除当前链表元素时，
    要修改map中下一个节点所对应的元素。
*/

import java.util.*;

class ListNode {
    public int val;
    public int key;
    public ListNode next;
    public ListNode(int key, int val) {
        this.key = key;
        this.val = val;
    }
}

public class LRUCache {

    private Map<Integer, ListNode> map;
    private ListNode head;
    private ListNode tail;
    private int size;
    private int capacity;

    public LRUCache(int capacity) {
        this.map = new HashMap<>();
        this.head = new ListNode(0, 0);
        this.tail = head;
        this.capacity = capacity;
        this.size = 0;
    }

    public int get(int key) {
        if (!this.map.containsKey(key)) return -1;
        ListNode preNode = this.map.get(key);
        ListNode curNode = preNode.next;
        if (this.tail != curNode){
            this.map.put(curNode.next.key, this.map.get(curNode.key));   // pay attention
            this.map.put(curNode.key, this.tail);                        // pay attention
            preNode.next = preNode.next.next;
            curNode.next = null;

            this.tail.next = curNode;
            this.tail = this.tail.next;
        }
        return curNode.val;
    }

    public void put(int key, int value) {
        if (this.map.containsKey(key)) {
            get(key);
            this.tail.val = value;
        }
        else {
            if (this.size < this.capacity) {
                appendNode(key, value);
            }
            else {
                ListNode removeNode = this.head.next;
                this.map.remove(removeNode.key);
                this.head.next = this.head.next.next;
                this.size--;
                if (removeNode.next != null){                            // pay attention
                    this.map.put(removeNode.next.key, this.head);
                }
                if (removeNode == this.tail)
                    this.tail = this.head;
                appendNode(key, value);
            }
        }
    }

    private void appendNode(int key, int value) {
        this.map.put(key, this.tail);
        this.tail.next = new ListNode(key, value);
        this.tail = this.tail.next;
        this.size++;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: round2/经典题目/Cache/LRUCache.java
/*
    leetcode 146
    注意由于是双向链表，删除一个节点时，这个节点，pre，next都要跟着变化
*/

import java.util.*;

class LRUCache {

    class Node {
        int key, val;
        Node pre, next;
        public Node(int key, int val) {
            this.key = key;
            this.val = val;
        }
    }

    private Map<Integer, Node> map;
    private Node head, tail;
    private int capacity, size;

    public LRUCache(int capacity) {
        map = new HashMap<>();
        head = new Node(0, 0);
        tail = head;
        this.capacity = capacity;
        size = 0;
    }

    public int get(int key) {
        if (!map.containsKey(key))
            return -1;

        Node currNode = map.get(key);
        if (currNode != tail) {
            currNode.pre.next = currNode.next;
            if (currNode.next != null)
                currNode.next.pre = currNode.pre;
            currNode.next = null;
            currNode.pre = tail;
            tail.next = currNode;
            tail = tail.next;
        }
        return currNode.val;
    }

    public void put(int key, int value) {
        if (map.containsKey(key)) {
            get(key);
            map.get(key).val = value;
        }
        else {
            if (size < capacity) {
                size++;
                Node currNode = new Node(key, value);
                map.put(key, currNode);
                tail.next = currNode;
                currNode.pre = tail;
                tail = tail.next;
            }
            else {
                Node removeNode = head.next;
                head.next = head.next.next;
                if (head.next != null)
                    head.next.pre = head;
                size--;
                map.remove(removeNode.key);
                put(key, value);
            }
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
